File: algorithm/200.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>

void dfs(char **grid, int gridSize, int gridColSize, int **visited, int x, int y)
{
    if (x < 0 || x >= gridSize)
    {
        return;
    }
    if (y < 0 || y >= gridColSize)
    {
        return;
    }
    if (grid[x][y] == '0' || visited[x][y] == 1)
    {
        return;
    }
    visited[x][y] = 1;
    dfs(grid, gridSize, gridColSize, visited, x - 1, y);
    dfs(grid, gridSize, gridColSize, visited, x + 1, y);
    dfs(grid, gridSize, gridColSize, visited, x, y - 1);
    dfs(grid, gridSize, gridColSize, visited, x, y + 1);
}

int numIslands(char **grid, int gridSize, int *gridColSize)
{
    int m = gridSize;
    int n = gridColSize[0];
    int **visited = (int **)malloc(m * sizeof(int *));
    for (int i = 0; i < m; i++)
    {
        visited[i] = (int *)calloc(n, sizeof(int));
    }
    int count = 0;
    for (int i = 0; i < m; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (grid[i][j] != '0' && visited[i][j] != 1)
            {
                dfs(grid, m, n, visited, i, j);
                count++;
            }
        }
    }
    return count;
}

File: go/goalgor/algor/200.go
var direction = [][]int{{1, 0}, {-1, 0}, {0, -1}, {0, 1}}

func numIslands(grid [][]byte) int {
	m := len(grid)
	n := len(grid[0])
	visited := make([][]int, m)
	for i := 0; i < m; i++ {
		visited[i] = make([]int, n)
	}
	var dfs func(int, int, [][]byte, [][]int)
	dfs = func(x, y int, grid [][]byte, visited [][]int) {
		grid[x][y] = '0'
		visited[x][y] = 1
		for i := 0; i < len(direction); i++ {
			newx := x + direction[i][0]
			newy := y + direction[i][1]
			if newx < 0 || newx >= m || newy < 0 || newy >= n || visited[newx][newy] == 1 {
				continue
			}
			if grid[newx][newy] == '1' {
				dfs(newx, newy, grid, visited)
			}
		}
	}
	count := 0
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' {
				dfs(i, j, grid, visited)
				count = count + 1
			}
		}
	}
	return count
}
